realms: Create default system user groups for internal realm.

sahil839 opened the pull request:
Since we include internal realms while creating system groups
in "0382_create_role_based_system_groups.py", we should do it
when creating new internal realms as well to be consistent.
This PR also modifies `bulk_create_users` to add the users
to the respective system groups.

And we would also need a migration to add the groups to internal
realms created after the previous migration.

<!-- Describe your pull request here.-->

 <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

sahil839 commented:
@timabbott this is ready for your review.

sahil839 commented:
@timabbott can you review this? Would be good to get this one completed as I will open the next PR for adding a new setting based on groups model tommorrow, and would be nice to get this one off the list.

timabbott commented:
This looks to me to be a correct implementation of fixing this inconsistency by adding these groups to the system bot realm.

I feel like it might be more correct to instead remove them from the system bot realm, if that realm doesn't contain users beyond the system bots... but I'm not sure how to code that in a safe fashion.

@mateuszmandera thoughts? I can see going either way on this, given our eventual goal to eliminate the system bot realm as a concept...

sahil839 commented:
We currently define internal bots in `settings.INTERNAL_BOTS` so we can probably check through emails I think. Not think there would be many of them with users other than system bots.

mateuszmandera commented:
> I feel like it might be more correct to instead remove them from the system bot realm, if that realm doesn't contain users beyond the system bots... but I'm not sure how to code that in a safe fashion.

Hmm, what would the advantage be? Isn't it simpler to just have the clear property `the system realm has system user groups` rather than it being conditioned on whether the realm has additional human users or not, with all the extra code that would be needed for ensuring that conditionality?

mateuszmandera commented on zerver/lib/bulk_create.py:
```diff
@@ -116,6 +118,26 @@ def bulk_create_users(
 
     Subscription.objects.bulk_create(subscriptions_to_create)
 
+    full_members_system_group = UserGroup.objects.get(
```
This codepath is also called via `do_import_system_bots` calling `create_users`, which calls this `bulk_create_users` - in `import_realm.py` - have you checked if this will work fine? E.g. if importing from another server, which has these `UserGroupMembership`s in the internal realm, won't an error happen due to the `UserGroupMembership`s being attempted to be created twice - once through this codepath and once through the import of user groups?

I haven't looked at this in-depth, just an edge case that came to mind that is not clear to me if was addressed - we can audit further if needed

sahil839 replied:
I checked it once again now and this will work fine. `do_import_system_bots` is called after importing realm, and we also check whether the bots are already created or not and I think we can assume that the `UserGroupMemberships` objects in the import data will only be present if `UserProfile` objects are present in import data.

mateuszmandera replied:
Okay, cool thanks for checking!

timabbott replied:
I think the comment here is incorrect right now, though, since this is called from the data import code path. However, as noted in the thread I started here:

https://chat.zulip.org/#narrow/stream/3-backend/topic/do_import_system_bots/near/1417790

I think that the data import code path is a noop (and would be a bug if it weren't) and should be deleted, so the comment is likely fine as written.

mateuszmandera commented on zerver/lib/bulk_create.py:
```diff
@@ -116,6 +118,26 @@ def bulk_create_users(
     Subscription.objects.bulk_create(subscriptions_to_create)
 
+    full_members_system_group = UserGroup.objects.get(
+        name="@role:fullmembers", realm=realm, is_system_group=True
```
Should these `"@role:fullmembers"` etc. string be deduplicated and instead be centrally  hard-coded in `UserGroup.FULL_MEMBERS` etc. or something like that? There's some other places where these strings are repeated. Or is that intended since maybe this popped up in the reviews of the original implementation?

sahil839 replied:
I don't think we discussed in the original implementation, but I liked the idea of deduplicating this, does `UserGroup.FULL_MEMBERS_GROUP_NAME` sounds good. For other cases we can probably use `UserGroup.SYSTEM_USER_GROUP_ROLE_MAP`, I know this would be long but we already declare name of user groups there.

mateuszmandera replied:
Yeah, that name sounds good to me

timabbott replied:
Yeah those names sound great.

timabbott replied:
Reading this, I think the `UserGroup.SYSTEM_USER_GROUP_ROLE_MAP` references feel cumbersome. I think it's pretty reasonable to define constants like `EVERYONE_ON_INTERNET_GROUP_NAME = "@role:internet"` for all 5-6 of them, and have `SYSTEM_USER_GROUP_ROLE_MAP` use those values in its definitions, rather than this approach.

I'll drop the last commit and also tweak the other two final commits to fit that proposed pattern.

mateuszmandera commented on zerver/lib/bulk_create.py:
```diff
@@ -116,6 +118,26 @@ def bulk_create_users(
+    )
+    group_memberships_to_create: List[UserGroupMembership] = []
+    for user_profile in profiles_to_create:
+        # All users are members since this function is only used to create bots
+        # and test and development environment users.
```
Can we add an `assert` here about `user_profile.role` so that we don't get broken group memberships state if this function is ever called with this assumption broken?

mateuszmandera commented on zerver/migrations/0401_create_role_based_groups_for_internal_realms.py:
```diff
@@ -0,0 +1,159 @@
+from django.utils.timezone import now as timezone_now
+
+
+def create_role_based_system_groups_for_internal_realms(
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
```
This migration is just basically just a copy of `zerver/migrations/0382_create_role_based_system_groups.py` run for the system bot realm, right? Seems worth it to note that in a comment somewhere in the file

mateuszmandera reviewed (commented):
Looks good I think with those couple of tiny tweaks mentioned

sahil839 commented:
Added the assert statement and a comment in migration as asked above.

mateuszmandera reviewed (approved):
Awesome, thanks!

sahil839 commented:
Added commits to add constants for user group names wherever requried and using them and `SYSTEM_USER_GROUP_ROLE_MAP` to get user groups from names and in tests.

timabbott commented:
This is great, merged with a few changes:
* The tweaks mentioned here: https://github.com/zulip/zulip/pull/22352#discussion_r941909110
* Made sure the comment for the migration appears at the top of the file; that's more likely to be visible when looking for the information. I left a copy at the bottom as well.

sahil839 commented:
@timabbott  This was not merge automatically because CI was failing due to linter error. I fixed it and also added commits for other group names. CI is not passing due to puppeteer error, which I think would be fixed by #22692.

sahil839 commented:
Ohh, this got auto-merged as I repushed again for CI to pass.

timabbott commented:
Yeah, that's a possibly surprising interaction with how GitHub's "merge once CI passes" feature works -- basically it'll merge the next time the PR passes CI, regardless of whether there are actual changes or not.

In this case, I've reviewed the new commits and they look great, so not a problem :)
